File: src/runner/fire.ts
import { HerdrError } from "../adapter/rpc";
import { fillAgentArgv } from "../config";
import { HERDR_METHOD_BY_NAME } from "../herdr-methods.generated";
import { PANE_READ_LINES, PANE_READ_SOURCE } from "../pane-read";
import type { FlatStep, PlaceholderValues } from "../workflows/types";
import { substitute, substituteParams } from "../workflows/substitute";
import { AGENT_NAME_RE } from "../workflows/inputs";
import { waitAgentDone } from "./agent-wait";
import { placeCommand } from "./place";
import { shellArgv } from "./shell";
import type { RunnerDeps, StepRunOptions } from "./types";

export type FireOutcome = {
  failed?: { ok: false; error: string };
  text?: string;
  bindings?: Record<string, string>;
  result?: Record<string, unknown>;
};

const INVOKING_AGENT = "{agent}";

export async function fail(
  deps: RunnerDeps,
  workflow: string,
  step: number,
  detail: string,
): Promise<string> {
  const text = `step ${step}: ${detail}`;
  const body = text.length > 500 ? `…${text.slice(-500)}` : text;
  await deps.notificationShow(`herdr-workflows: ${workflow} failed`, body).catch(() => undefined);
  return body;
}

function autofill(
  method: string,
  params: Record<string, unknown> | undefined,
  ctx: StepRunOptions["ctx"],
): Record<string, unknown> {
  const out = { ...params };
  const props = HERDR_METHOD_BY_NAME.get(method)?.params.properties ?? {};
  if (out.pane_id === undefined && ctx.paneId && props.pane_id) out.pane_id = ctx.paneId;
  if (out.tab_id === undefined && ctx.tabId && props.tab_id) out.tab_id = ctx.tabId;
  if (out.workspace_id === undefined && ctx.workspaceId && props.workspace_id) {
    out.workspace_id = ctx.workspaceId;
  }
  return out;
}

function resolveAgentName(stepName: string, values: PlaceholderValues): string {
  if (stepName === INVOKING_AGENT) return values.agent ?? "";
  const m = AGENT_NAME_RE.exec(stepName);
  return m ? (values[m[1]!] ?? "") : stepName;
}

function commandArgv(
  step: FlatStep & { action: { kind: "run" } },
  values: PlaceholderValues,
): string[] {
  const payload = step.action.payload;
  if (payload.form === "argv") return payload.argv.map((el) => substitute(el, values));
  return shellArgv(payload.command, payload.shell);
}

function readPath(obj: unknown, path: string): unknown {
  let cur: unknown = obj;
  for (const part of path.split(".")) {
    if (cur === null || cur === undefined || typeof cur !== "object") return undefined;
    cur = (cur as Record<string, unknown>)[part];
  }
  return cur;
}

function bindOutMap(
  fields: Record<string, string>,
  result: Record<string, unknown>,
): { ok: true; bindings: Record<string, string> } | { ok: false; error: string } {
  const bindings: Record<string, string> = {};
  for (const [name, path] of Object.entries(fields)) {
    const value = readPath(result, path);
    if (value === undefined) {
      const type = typeof result.type === "string" ? result.type : "unknown";
      return { ok: false, error: `out.${name}: path '${path}' missing on result.type '${type}'` };
    }
    bindings[name] = typeof value === "string" ? value : JSON.stringify(value);
  }
  return { ok: true, bindings };
}

function substituteEnv(
  env: Record<string, string> | undefined,
  values: PlaceholderValues,
): Record<string, string> | undefined {
  if (!env) return undefined;
  return Object.fromEntries(Object.entries(env).map(([k, v]) => [k, substitute(v, values)]));
}

function resolveRunEnv(
  step: FlatStep & { action: { kind: "run" } },
  opts: StepRunOptions,
  values: PlaceholderValues,
  env: NodeJS.ProcessEnv,
): { cwd: string; mergedEnv: NodeJS.ProcessEnv; stepEnv?: Record<string, string> } {
  const cwd = step.action.cwd !== undefined ? substitute(step.action.cwd, values) : opts.ctx.cwd;
  const stepEnv = substituteEnv(step.action.env, values);
  return { cwd, stepEnv, mergedEnv: { ...env, ...stepEnv } };
}

async function withOutMap(
  opts: StepRunOptions,
  n: number,
  out: FlatStep["out"],
  result: Record<string, unknown>,
): Promise<FireOutcome> {
  if (out?.kind !== "map") return { result };
  const bound = bindOutMap(out.fields, result);
  if (!bound.ok) {
    return { failed: { ok: false, error: await fail(opts.deps, opts.name, n, bound.error) } };
  }
  return { bindings: bound.bindings, result };
}

async function fireRun(
  opts: StepRunOptions,
  step: FlatStep & { action: { kind: "run" } },
  values: PlaceholderValues,
  n: number,
  env: NodeJS.ProcessEnv,
): Promise<FireOutcome> {
  const place = step.action.in;
  const { cwd, mergedEnv, stepEnv } = resolveRunEnv(step, opts, values, env);

  if (place === "here") {
    const payload = step.action.payload;
    const result =
      payload.form === "argv"
        ? await opts.deps.runArgv(
            payload.argv.map((el) => substitute(el, values)),
            { cwd, env: mergedEnv, timeoutMs: step.timeoutMs },
          )
        : await opts.deps.runShell(payload.command, {
            cwd,
            env: mergedEnv,
            timeoutMs: step.timeoutMs,
            shell: payload.shell,
          });
    if (result.stderr) opts.onStderr?.(result.stderr);
    if (!result.ok) {
      return {
        failed: {
          ok: false,
          error: await fail(opts.deps, opts.name, n, result.stderr.trim() || "nonzero exit"),
        },
      };
    }
    const bindings: Record<string, string> = {};
    if (step.out?.kind === "text") bindings[step.out.name] = result.stdout;
    return { text: result.stdout, bindings };
  }

  const label =
    step.name ??
    (step.action.payload.form === "argv"
      ? step.action.payload.argv[0] || "run"
      : step.action.payload.command.split(/\s+/)[0] || "run");
  const argv = commandArgv(step, values);
  const applied = await placeCommand(opts, place, argv, label, cwd, stepEnv, step.action.ratio);

  if (step.wait.kind === "detach") {
    return {
      bindings: step.out?.kind === "map" ? undefined : undefined,
      result: { tab_id: applied.tabId, pane_id: applied.paneId, workspace_id: applied.workspaceId },
    };
  }
  if (step.wait.kind === "regex") {
    await opts.deps.waitOutput(applied.paneId, step.wait.pattern, step.timeoutMs ?? 60_000);
  } else {
    // block: no pane process-wait API; treat placement success as completion unless regex/detach
  }

  return withOutMap(opts, n, step.out, {
    tab_id: applied.tabId,
    pane_id: applied.paneId,
    workspace_id: applied.workspaceId,
    type: "layout",
    layout: {
      tab_id: applied.tabId,
      focused_pane_id: applied.paneId,
      workspace_id: applied.workspaceId,
    },
  });
}

async function fireAgent(
  opts: StepRunOptions,
  step: FlatStep & { action: { kind: "agent" } },
  values: PlaceholderValues,
  n: number,
): Promise<FireOutcome> {
  const name = resolveAgentName(step.action.agent, values);
  if (step.action.agent === INVOKING_AGENT && !name) {
    return {
      failed: {
        ok: false,
        error: await fail(
          opts.deps,
          opts.name,
          n,
          "invoking agent unresolved — run from agent pane",
        ),
      },
    };
  }
  const template = opts.agents[name];
  if (!template) {
    return {
      failed: {
        ok: false,
        error: await fail(opts.deps, opts.name, n, `unknown agent '${name}'`),
      },
    };
  }
  const prompt = step.action.prompt !== undefined ? substitute(step.action.prompt, values) : "";
  const cwd = step.action.cwd !== undefined ? substitute(step.action.cwd, values) : opts.ctx.cwd;
  const stepEnv = substituteEnv(step.action.env, values);
  const place = step.action.in;
  const applied = await placeCommand(
    opts,
    place === "here" ? "tab" : place,
    fillAgentArgv(template, prompt),
    name,
    cwd,
    stepEnv,
    step.action.ratio,
  );

  if (step.wait.kind === "detach") {
    return { result: { tab_id: applied.tabId, pane_id: applied.paneId } };
  }
  if (step.wait.kind === "regex") {
    await opts.deps.waitOutput(applied.paneId, step.wait.pattern, step.timeoutMs ?? 1_800_000);
  } else {
    await waitAgentDone(applied.paneId, step.timeoutMs ?? 1_800_000, {
      agentStatus: opts.deps.agentStatus,
      sleep: opts.deps.sleep ?? ((ms) => Bun.sleep(ms)),
      now: opts.deps.now,
      pollMs: opts.deps.agentWaitPollMs,
      idleGraceMs: opts.deps.agentWaitIdleGraceMs,
      onBlocked: () =>
        opts.deps.notificationShow(
          `herdr-workflows: ${opts.name} waiting`,
          `agent blocked on step ${n} — needs your input`,
        ),
    });
  }
  const text = (
    await opts.deps.paneRead(applied.paneId, {
      source: PANE_READ_SOURCE,
      lines: PANE_READ_LINES,
    })
  ).trim();
  const bindings: Record<string, string> = {};
  if (step.out?.kind === "text") bindings[step.out.name] = text;
  return { text, bindings };
}

async function firePrimitive(
  opts: StepRunOptions,
  step: FlatStep & { action: { kind: "primitive" } },
  values: PlaceholderValues,
  n: number,
): Promise<FireOutcome> {
  const params = autofill(
    step.action.method,
    substituteParams(step.action.params, values),
    opts.ctx,
  );
  const result = (await opts.deps.herdrCall(step.action.method, params)) as Record<string, unknown>;
  return withOutMap(opts, n, step.out, result);
}

export async function fire(
  opts: StepRunOptions,
  step: FlatStep,
  values: PlaceholderValues,
  n: number,
  env: NodeJS.ProcessEnv,
): Promise<FireOutcome> {
  try {
    if (step.action.kind === "run") {
      return await fireRun(opts, step as FlatStep & { action: { kind: "run" } }, values, n, env);
    }
    if (step.action.kind === "agent") {
      return await fireAgent(opts, step as FlatStep & { action: { kind: "agent" } }, values, n);
    }
    if (step.action.kind === "primitive") {
      return await firePrimitive(
        opts,
        step as FlatStep & { action: { kind: "primitive" } },
        values,
        n,
      );
    }
    return {
      failed: {
        ok: false,
        error: await fail(opts.deps, opts.name, n, "internal: include must be handled by dispatch"),
      },
    };
  } catch (error) {
    const detail =
      error instanceof HerdrError || error instanceof Error ? error.message : String(error);
    return { failed: { ok: false, error: await fail(opts.deps, opts.name, n, detail) } };
  }
}

File: rust/src/runner/fire.rs
//! Pane-creating step dispatch (`open` / `agent` / `herdr`). Port of
//! `src/runner/fire.ts`, including the `fail` notifier shared with
//! `dispatch` and `runner`.

use serde_json::{Map, Value};

use super::agent_wait::wait_agent_done;
use super::context::{PANE_READ_LINES, PANE_READ_SOURCE};
use super::deps::{Herdr, StepResult, StepRunOptions};
use super::shell::shell_argv;
use crate::config::fill_agent_argv;
use crate::herdr::cli::{LayoutApplyParams, PaneReadOpts};
use crate::workflow::placeholder::agent_input_ref;
use crate::workflow::steps::FlatStep;
use crate::workflow::substitute::{PlaceholderValues, substitute, substitute_params};

const INVOKING_AGENT: &str = "{agent}";

/// `FireOutcome` — any subset of failed/last/tabId, as TS.
#[derive(Debug, Default)]
pub struct FireOutcome {
    pub failed: Option<StepResult>,
    pub last: Option<String>,
    pub tab_id: Option<String>,
}

/// `fail` — notify (best-effort) and return the error text, truncated to
/// the last 500 chars with an ellipsis prefix.
pub fn fail(herdr: &dyn Herdr, workflow: &str, step: usize, detail: &str) -> String {
    let text = format!("step {step}: {detail}");
    let body = if text.chars().count() > 500 {
        let tail: String = text.chars().skip(text.chars().count() - 500).collect();
        format!("…{tail}")
    } else {
        text
    };
    let _ = herdr.notification_show(&format!("herdr-workflows: {workflow} failed"), Some(&body));
    body
}

/// `autofill` — herdr-step params get the invoking ids when not pinned.
/// Input is the workflow's `BTreeMap` params; output is the socket payload
/// object.
fn autofill(
    params: Option<&std::collections::BTreeMap<String, Value>>,
    opts: &StepRunOptions,
) -> Map<String, Value> {
    let mut out: Map<String, Value> = params.cloned().unwrap_or_default().into_iter().collect();
    if !out.contains_key("pane_id")
        && let Some(pane_id) = &opts.ctx.pane_id
    {
        out.insert("pane_id".to_string(), Value::String(pane_id.clone()));
    }
    if !out.contains_key("tab_id")
        && let Some(tab_id) = &opts.ctx.tab_id
    {
        out.insert("tab_id".to_string(), Value::String(tab_id.clone()));
    }
    if !out.contains_key("workspace_id")
        && let Some(workspace_id) = &opts.ctx.workspace_id
    {
        out.insert(
            "workspace_id".to_string(),
            Value::String(workspace_id.clone()),
        );
    }
    out
}

/// `resolveAgentName` — `{agent}` from the invocation snapshot, `{input.*}`
/// from resolved inputs, anything else verbatim.
fn resolve_agent_name(step_name: &str, values: &PlaceholderValues) -> String {
    if step_name == INVOKING_AGENT {
        return values.agent.clone();
    }
    if let Some(input) = agent_input_ref(step_name) {
        return values.inputs.get(&input).cloned().unwrap_or_default();
    }
    step_name.to_string()
}

/// `fireOpen`.
fn fire_open(
    opts: &StepRunOptions,
    command: &str,
    wait_for: Option<&str>,
    timeout_ms: Option<u64>,
) -> Result<FireOutcome, String> {
    // `" x".split(/\s+/)[0]` is "" → TS falls back to "open"; mirror that.
    let first = command.split(char::is_whitespace).next().unwrap_or("");
    let label = if first.is_empty() { "open" } else { first };
    let applied = opts
        .herdr
        .layout_apply(&LayoutApplyParams {
            workspace_id: opts.ctx.workspace_id.clone(),
            tab_label: label.to_string(),
            tab_id: None,
            cwd: opts.ctx.cwd.clone(),
            command: shell_argv(command),
            label: label.to_string(),
            env: Default::default(),
            focus: Some(true),
        })
        .map_err(|e| e.message)?;
    if let Some(pattern) = wait_for {
        opts.herdr
            .wait_output(
                &applied.pane_id,
                pattern,
                timeout_ms.expect("wait_for implies timeout"),
            )
            .map_err(|e| e.message)?;
    }
    Ok(FireOutcome {
        tab_id: Some(applied.tab_id),
        ..FireOutcome::default()
    })
}

/// Borrowed view of a `FlatStep::Agent`'s fields, so `fire_agent` stays
/// under the argument-count lint.
struct AgentStep<'a> {
    name: &'a str,
    prompt: Option<&'a str>,
    wait: bool,
    timeout_ms: Option<u64>,
    close_source: bool,
}

/// `fireAgent`. Returns `Ok(None)` when the step already produced a
/// `failed` outcome (unresolved/unknown agent), like the TS early returns.
fn fire_agent(
    opts: &StepRunOptions,
    step: &AgentStep,
    values: &PlaceholderValues,
    n: usize,
    last: &str,
) -> Result<Option<FireOutcome>, String> {
    let resolved = resolve_agent_name(step.name, values);
    if step.name == INVOKING_AGENT && resolved.is_empty() {
        let error = fail(
            opts.herdr,
            opts.name,
            n,
            "invoking agent unresolved — run from agent pane",
        );
        return Ok(Some(FireOutcome {
            failed: Some(StepResult::failed(error, last.to_string())),
            ..FireOutcome::default()
        }));
    }
    let Some(template) = opts.agents.get(&resolved) else {
        let error = fail(
            opts.herdr,
            opts.name,
            n,
            &format!("unknown agent '{resolved}'"),
        );
        return Ok(Some(FireOutcome {
            failed: Some(StepResult::failed(error, last.to_string())),
            ..FireOutcome::default()
        }));
    };
    let prompt = step
        .prompt
        .map_or_else(String::new, |p| substitute(p, values));
    let applied = opts
        .herdr
        .layout_apply(&LayoutApplyParams {
            workspace_id: opts.ctx.workspace_id.clone(),
            tab_label: resolved.clone(),
            tab_id: None,
            cwd: opts.ctx.cwd.clone(),
            command: fill_agent_argv(template, &prompt),
            label: resolved,
            env: Default::default(),
            focus: Some(true),
        })
        .map_err(|e| e.message)?;
    // Close source only after target opened — failure above leaves the
    // original tab intact.
    if step.close_source
        && let Some(tab_id) = &opts.ctx.tab_id
    {
        opts.herdr.tab_close(tab_id).map_err(|e| e.message)?;
    }
    if !step.wait {
        return Ok(Some(FireOutcome {
            tab_id: Some(applied.tab_id),
            ..FireOutcome::default()
        }));
    }
    let mut status = |pane: &str| opts.herdr.agent_status(pane);
    let on_blocked = || {
        opts.herdr.notification_show(
            &format!("herdr-workflows: {} waiting", opts.name),
            Some(&format!("agent blocked on step {n} — needs your input")),
        )
    };
    wait_agent_done(
        &applied.pane_id,
        std::time::Duration::from_millis(step.timeout_ms.expect("wait implies timeout")),
        &mut status,
        opts.wait_clock,
        Some(&on_blocked),
    )
    .map_err(|e| e.message)?;
    let text = opts
        .herdr
        .pane_read(
            &applied.pane_id,
            PaneReadOpts {
                source: Some(PANE_READ_SOURCE),
                lines: Some(PANE_READ_LINES),
            },
        )
        .map_err(|e| e.message)?;
    Ok(Some(FireOutcome {
        last: Some(text.trim().to_string()),
        tab_id: Some(applied.tab_id),
        failed: None,
    }))
}

/// `fire` — non-shell steps. Any herder error becomes a `failed` outcome
/// carrying the notified message.
pub fn fire(
    opts: &StepRunOptions,
    step: &FlatStep,
    values: &PlaceholderValues,
    n: usize,
    last: &str,
) -> FireOutcome {
    let result: Result<Option<FireOutcome>, String> = match step {
        FlatStep::Open {
            command,
            wait_for,
            timeout_ms,
        } => fire_open(opts, command, wait_for.as_deref(), *timeout_ms).map(Some),
        FlatStep::Agent {
            name,
            prompt,
            wait,
            timeout_ms,
            close_source,
        } => fire_agent(
            opts,
            &AgentStep {
                name,
                prompt: prompt.as_deref(),
                wait: *wait,
                timeout_ms: *timeout_ms,
                close_source: *close_source,
            },
            values,
            n,
            last,
        ),
        FlatStep::Herdr { method, params } => opts
            .herdr
            .herdr_call(
                method,
                autofill(substitute_params(params.as_ref(), values).as_ref(), opts),
            )
            .map(|_| None)
            .map_err(|e| e.message),
        FlatStep::Shell { .. } => unreachable!("shell steps never reach fire"),
    };
    match result {
        Ok(Some(outcome)) => outcome,
        Ok(None) => FireOutcome::default(),
        Err(detail) => FireOutcome {
            failed: Some(StepResult::failed(
                fail(opts.herdr, opts.name, n, &detail),
                last.to_string(),
            )),
            ..FireOutcome::default()
        },
    }
}
